dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,425 @@
+
+#### BooleanAttributePart.\_$setValue
+
+`BooleanAttributePart` also extends `AttributePart` and overrides `_commitValue`. When a value is committed to a `BooleanAttributePart`, a falsy or `nothing` value will remove the attribute, and a truthy value will set the attribute with an empty string attribute value.
+
+#### EventPart.\_$setValue
+
+`EventPart` also extends `AttributePart`, and overrides `_$setValue` to take the user provided event listener and then call `this.element.addEventListener(attributeName, value)`.
+The EventPart ensures that event listeners are added and cleaned up between renders and if the passed-in value changes.
+
+A value of `null`, `undefined`, or `nothing` will remove any previously set listeners without adding a new listener.
+
+#### ElementPart.\_$setValue
+
+Nothing can be committed via an `ElementPart`. Instead `resolveDirective` is called, allowing directives to hook into the `ElementPart` position.
+
+For example, [`@lit-labs/motion` has an `animate` directive](https://lit.dev/playground/#sample=examples/motion-simple) that can be applied on an element in the `ElementPart` location to then manage the elements animations.
+
+# Efficient Updates
+
+So far we've detailed what happens on the first render: `render(counterUi(0), container)` in the example, but what happens when the "Increment" button is clicked and `render(counterUi(1), container)` is invoked?
+
+The `render()` function looks for a `'_$litPart$'` property on the container and finds a `ChildPart`. `_$setValue` is called on the container's `ChildPart` passing the `TemplateResult` returned from `counterUi(1)`.
+
+The `TemplateResult` for `counterUi(1)` is:
+
+```js
+{
+  _$litType$: 1,
+  strings: [
+    '<span class="',
+    '">',
+    '</span><button @click=',
+    '>Increment</button>'
+  ],
+  values: [
+    "odd",
+    1,
+    () => render(counterUi(2), container)
+  ]
+}
+```
+
+Instead of redoing the work in the **Prepare** phase, the Lit `Template` is fetched from the cache using the template's `strings` Template Strings Array.
+
+> **Note**
+> The reference to the Template Strings Array is the cache key, not the value of the key. This is an O(1) lookup and not changed by the length of the static contents.
+
+The container's `ChildPart` stores the last committed value - which in this case is a `TemplateInstance` referencing the `Template` returned from the cache. Because we're rendering the same `Template` to this `ChildPart` the **Create** phase is skipped.
+
+All that is left is to iterate over the previously created `Part`s and call `_$setValue` with the new values. For demonstrative purposes, inlined this looks like:
+
+```js
+AttributePart._$setValue('odd');
+ChildPart._$setValue(1);
+EventPart._$setValue(() => render(counterUi(2), container));
+```
+
+This results in the DOM updating with a green underlined number one, and the event listener is swapped with the new listener.
```

Comment:
Not super necessary but it would've also been nice to be able to callout if any of the values being set here were the same, we don't actually do any DOM operations. That would require some piece of the original expression in the example not being changed by the click.

It's not like we're really describing what happens at each of the `_$setValue()` here anyway so it seems fine to just leave it there.

Feel free to ignore this!

Reply:
Added a sentence: "If any of the values being set remain the same (by triple equal check), Lit will not do any DOM
operations." Because it's nice to mention! Good comment!